packages/config-helpers/src/define-config.js

```diff
@@ -0,0 +1,380 @@
+
+		// if it's an array then traverse it as well
+		if (Array.isArray(configList[i])) {
+			yield* flatTraverse(configList[i], newIndexPath);
+			continue;
+		}
+
+		yield { indexPath: newIndexPath, value: configList[i] };
+	}
+}
+
+/**
+ * Extends a list of config files by creating every combination of base and extension files.
+ * @param {(string|string[])[]} [baseFiles] The base files.
+ * @param {(string|string[])[]} [extensionFiles] The extension files.
+ * @return {(string|string[])[]} The extended files.
+ */
+function extendConfigFiles(baseFiles = [], extensionFiles = []) {
+	if (!extensionFiles.length) {
+		return baseFiles.concat();
+	}
+
+	if (!baseFiles.length) {
+		return extensionFiles.concat();
+	}
+
+	/** @type {(string|string[])[]} */
+	const result = [];
+
+	for (const baseFile of baseFiles) {
+		for (const extensionFile of extensionFiles) {
+			/*
+			 * Each entry can be a string or array of strings. The end result
+			 * needs to be an array of strings, so we need to be sure to include
+			 * all of the items when there's an array.
+			 */
+
+			const entry = [];
+
+			if (Array.isArray(baseFile)) {
+				entry.push(...baseFile);
+			} else {
+				entry.push(baseFile);
+			}
+
+			if (Array.isArray(extensionFile)) {
+				entry.push(...extensionFile);
+			} else {
+				entry.push(extensionFile);
+			}
+
+			result.push(entry);
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Extends a config object with another config object.
+ * @param {Config} baseConfig The base config object.
+ * @param {string} baseConfigName The name of the base config object.
+ * @param {Config} extension The extension config object.
+ * @param {string} extensionName The index of the extension config object.
+ * @return {Config} The extended config object.
+ */
+function extendConfig(baseConfig, baseConfigName, extension, extensionName) {
+	const result = { ...extension };
+
+	// for files we need to create every combination of base and extension files
+	if (baseConfig.files) {
+		result.files = extendConfigFiles(baseConfig.files, extension.files);
+	}
+
+	// for ignores we just concatenation the extension ignores onto the base ignores
+	if (baseConfig.ignores) {
+		result.ignores = baseConfig.ignores.concat(extension.ignores ?? []);
+	}
+
+	result.name = `${baseConfigName} > ${extensionName}`;
+
+	return result;
+}
+
+/**
+ * Processes a list of extends elements.
+ * @param {ConfigWithExtends} config The config object.
+ * @param {WeakMap<Config, string>} configNames The map of config objects to their names.
+ * @return {Config[]} The flattened list of config objects.
+ */
+function processExtends(config, configNames) {
+	if (!config.extends) {
+		return [config];
+	}
+
+	const {
+		/** @type {Config[]} */
+		extends: extendsList,
+
+		/** @type {Config} */
+		...configObject
+	} = config;
+
+	const extensionNames = new WeakMap();
+
+	// replace strings with the actual configs
+	const objectExtends = extendsList.map(extendsElement => {
```

Comment:
I think it would be good to validate whether the value of `extends` is an array, and throw a descriptive error if it isn't. Currently, for example,  `defineConfig({extends: {}})` fails with generic `TypeError: extendsList.map is not a function`.